### DISCUSSION TITLE
Task - Spring Boot - Spring Security - Protecting the Treasure

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#80] [Tasks]
# Indiana Jones and the Temple of Doom

Create a Spring Boot application called `temple-of-doom`.

Create a resource named `TreasureController` which has three methods: `welcome()`, `findTreasure(String secretCode)`, and  `createTreasure()`.

The `welcome()` method returns a message string `"Welcome treasure hunter."`. 
The `findTreasure(String secretCode)` method returns `"Congratulations! You have discovered the secret treasure."`, if the string argument passed to the method is `"open sesame"`. Else, it returns `"You are doomed!"` The secretCode is passed by a _query parameter_.
The `createTreasure()` method returns the String `"A new hidden treasure has been created"`

Add Spring Security to the application and guard the application against unauthorized access as follows:
- Any visitor can visit the temple of doom.
- Configure the app such that only the "USER" **"indiana_jones"** with the password **"open sesame"** can find the hidden treasure.
- Configure the app so that only an "ALIEN" with the username **"superman"** and password **"kryptonite"** can create a new treasure. 

You can refer to the demo Spring Security authentication and authorization app [here](https://github.com/akash-coded/spring-framework/tree/main/eclipse-projects/spring-boot/spring-boot-security-auth-demo-app).

**Comment by thiripurasundari1025** [comment]
# Security Configuration
````java
@EnableWebSecurity
public class SpringSecurityConfiguration extends WebSecurityConfigurerAdapter{
	
	@Override
	protected AuthenticationManager authenticationManager() throws Exception {
		
		return super.authenticationManager();
	}

	@Override
	protected void configure(AuthenticationManagerBuilder auth) throws Exception {
	auth.inMemoryAuthentication()
	.withUser("indiana_jones")
	.password("open sesame")
	.roles("USER")
	.and()
	.withUser("superman")
	.password("kryptonite")
	.roles("ALIEN");
		
	}
	
	@Bean
	public PasswordEncoder getPasswordEncoder() {
		return NoOpPasswordEncoder.getInstance();
		
	}
	
	protected void configure(HttpSecurity http) throws Exception{
		http.authorizeRequests()
		.antMatchers("/create").hasRole("ALIEN")
		.antMatchers("/find/{secretCode}").hasRole("USER")
		.antMatchers("/").permitAll()
		.and().formLogin();
		
	}
}

````

# TreasureController
````java
@RestController
public class TreasureController {

	@GetMapping("/")
	private String welcome() {
		return "Welcome treasure hunter.";
	}

	@GetMapping("/find/{secretCode}")
	private String findTreasure(@PathVariable String secretCode) {
		System.out.println("jh"+secretCode);
		if (secretCode.equals("open sesame"))
			return "Congratulations! You have discovered the secret treasure.";
		else
			return "You are doomed!";
	}
	@GetMapping("/create")
	private String createTreasure(){
		return  "A new hidden treasure has been created";
		
	}
	
}

````
 # Application.properties
````java
server.port=8081

````

<img width="577" alt="Screenshot_36" src="https://user-images.githubusercontent.com/105277740/168591087-3bbab3cd-4a46-46b3-81a3-4d783d2a027f.png">

<img width="693" alt="Screenshot_33" src="https://user-images.githubusercontent.com/105277740/168590763-968be126-4da8-40cd-9837-d5f3602731ae.png">
<img width="705" alt="Screenshot_34" src="https://user-images.githubusercontent.com/105277740/168590776-780ab9c6-d01c-4fe4-a929-6faaf3069974.png">
<img width="817" alt="Screenshot_37" src="https://user-images.githubusercontent.com/105277740/168591368-757bf8c0-9dec-49b8-a2df-4189e51df618.png">

<img width="588" alt="Screenshot_35" src="https://user-images.githubusercontent.com/105277740/168590782-7af97ca4-ec79-4806-8182-0ff29d3592b9.png">

**Comment by KalaiselviRajavadivel** [comment]
# TreasureController

```java
package com.spring.security.demo.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TreasureController {
	@RequestMapping("/")
	public String home() {
		return"Welcome treasure hunter";
	}
	
	@GetMapping("findTreasure/{secretCode}")
	public String findTreasure(@PathVariable String secretCode) {
		if(secretCode.equals("open sesame")) {
		return "Congratulations! You have discovered the secret treasure.";
	}
		else {
			return "You are doomed!";
		}
	}
	@GetMapping("/createTreasure")
	public String createTreasure() {
		return"A new hidden treasure has been created";
	}
}

```
# Spring Security Configuration

```java
import org.springframework.context.annotation.Bean;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.crypto.password.NoOpPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

@SuppressWarnings("deprecation")
@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

	@Override
	protected void configure(AuthenticationManagerBuilder auth) throws Exception {
		// Set your configuration on the auth object	
		auth.inMemoryAuthentication()
				.withUser("indiana_jones")
				.password("open sesame")
				.roles("USER")
				.and()
				.withUser("superman")
				.password("kryptonite")
				.roles("ALIEN");
	}
	
	@Bean
	public PasswordEncoder getPasswordEncoder() {
		return NoOpPasswordEncoder.getInstance();
	}
	
	protected void configure(HttpSecurity http) throws Exception {
		http.authorizeRequests()
				.antMatchers("/createTreasure").hasRole("ALIEN")
				.antMatchers("/findTreasure/{secretCode}").hasAnyRole("USER")
				.antMatchers("/").permitAll()
				.and().formLogin();
	}
 	
}
```

<img width="841" alt="snap3" src="https://user-images.githubusercontent.com/105358230/168577861-39814ebb-e6b6-4af9-a8e7-f985606e3b58.PNG">
<img width="591" alt="snap4" src="https://user-images.githubusercontent.com/105358230/168577871-28c1491d-3848-4571-a99e-3c4a21ee0754.PNG">
<img width="581" alt="snap5" src="https://user-images.githubusercontent.com/105358230/168577879-ad39ef81-afe7-4784-8874-dae76627aadc.PNG">
<img width="767" alt="snap1" src="https://user-images.githubusercontent.com/105358230/168595716-dd7990ab-9884-489e-a3cc-92797130643c.PNG">
<img width="613" alt="snap3" src="https://user-images.githubusercontent.com/105358230/168595753-57597681-88b5-496f-b349-b99268a2c414.PNG">

<img width="768" alt="snap2" src="https://user-images.githubusercontent.com/105358230/168595741-8b681e95-f828-4de0-ac70-f81194e4efcd.PNG">

**Comment by GouthamGunji** [comment]
# Treasure Controller
```java
package com.spring.boot.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TreasureController {

    @GetMapping("/welcome")
    public String welcome() {
        return "Welcome treasure hunter.";
    }

    @GetMapping("/findtreasure")
    public String findTreasure(@RequestParam String key) {
        if (key.equals("open sasame"))
            return "Congratulations! You have discovered the secret treasure.";
        else
            return "You are doomed!";
    }

    @GetMapping("/create")
    public String createTreasure(){
        return  "A new hidden treasure has been created";

    }

}
```

# SecurityConfiguration
```java
package com.spring.boot.configuration;

import org.springframework.context.annotation.Bean;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.crypto.password.NoOpPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;

@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    @Override
    protected AuthenticationManager authenticationManager() throws Exception {

        return super.authenticationManager();
    }

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        // Set your configuration on the auth object
        auth.inMemoryAuthentication()
                .withUser("indiana_jones")
                .password("open sesame")
                .roles("USER")
                .and()
                .withUser("superman")
                .password("kryptonite")
                .roles("ALIEN");
    }

    @Bean
    public PasswordEncoder getPasswordEncoder() {
        return NoOpPasswordEncoder.getInstance();
    }

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .antMatchers("/createTreasure").hasRole("ALIEN")
                .antMatchers("/findTreasure").hasAnyRole("USER")
                .antMatchers("/welcome").permitAll()
                .and().formLogin();
    }

}
	
```
# application.properties

server.port=8081



# Screenshots

<img width="317" alt="image" src="https://user-images.githubusercontent.com/61229602/168610026-d6d3073b-d56b-4bc1-a52c-2d9ef02cbd0c.png">

<img width="707" alt="image" src="https://user-images.githubusercontent.com/61229602/168611303-8a231f47-7441-4104-b603-69f07b38f944.png">


<img width="404" alt="image" src="https://user-images.githubusercontent.com/61229602/168610584-ac65a154-8285-463d-b819-cff710f67496.png">


<img width="345" alt="image" src="https://user-images.githubusercontent.com/61229602/168610688-c1201ff9-a0fd-411b-8fc2-32f6fda0c08c.png">

<img width="788" alt="image" src="https://user-images.githubusercontent.com/61229602/168611154-824c8de2-5a2d-495b-8d56-d8c234a16f7a.png">


<img width="312" alt="image" src="https://user-images.githubusercontent.com/61229602/168610890-ef185cf7-2508-4f6d-9827-328600d95e3a.png">

**Comment by Anandhi22** [comment]
## Treasure Controller
```java
package springbootsecurityapp.demo.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class TreasureController {

	@GetMapping("/")
	private String welcome() {
		return "Welcome treasure hunter.";
	}
	@GetMapping("/find")
	private String findTreasure(String secretCode) {
		return "Congratulations! You have discovered the secret treasure.";
	}
}

```

## Application Properties
server.port = 8085
spring.security.user.name=indiana_jones
spring.security.user.password=open sesame

## Screenshot
![Uploading Screenshot (74).png…]()

**Comment by AgeshAalam** [comment]
# Controller
```Java

@RestController
public class TreasureController {

    @GetMapping("/welcome")
    public String welcome() {
        return "Welcome treasure hunter.";
    }

    @GetMapping("/findtreasure")
    public String findTreasure(@RequestParam String key) {
        if (key.equals("open sasame"))
            return "Congratulations! You have discovered the secret treasure.";
        else
            return "You are doomed!";
    }

    @GetMapping("/create")
    public String createTreasure(){
        return  "A new hidden treasure has been created";

    }

}
```
# SecurityConfiguration

```Java

@EnableWebSecurity
public class SecurityConfiguration extends WebSecurityConfigurerAdapter {

    @Override
    protected AuthenticationManager authenticationManager() throws Exception {

        return super.authenticationManager();
    }

    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        // Set your configuration on the auth object
        auth.inMemoryAuthentication()
                .withUser("indiana_jones")
                .password("open sesame")
                .roles("USER")
                .and()
                .withUser("superman")
                .password("kryptonite")
                .roles("ALIEN");
    }

    @Bean
    public PasswordEncoder getPasswordEncoder() {
        return NoOpPasswordEncoder.getInstance();
    }

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .antMatchers("/createTreasure").hasRole("ALIEN")
                .antMatchers("/findTreasure").hasAnyRole("USER")
                .antMatchers("/welcome").permitAll()
                .and().formLogin();
    }

}
```
# Application properties

```Java
server.port=9090

```
![image](https://user-images.githubusercontent.com/44727659/168735250-db93db29-e429-43bf-a141-a4af1d0fcbe4.png)

![image](https://user-images.githubusercontent.com/44727659/168736177-1b85699b-5182-47e9-9a00-362963c22676.png)

![image](https://user-images.githubusercontent.com/44727659/168736955-69216544-deff-463a-b3f8-762199c9475e.png)

![image](https://user-images.githubusercontent.com/44727659/168737247-e36ea143-1ffe-471e-ab00-0b8788517633.png)

**Comment by Svishva** [comment]
##### VisweshwaranM


## SecurityConfigurations .java
```java
@EnableWebSecurity
public class SecurityConfigurations extends WebSecurityConfigurerAdapter {

	@Override
	protected void configure(AuthenticationManagerBuilder authenticationManagerBuilder) throws Exception {

		authenticationManagerBuilder.inMemoryAuthentication().withUser("indiana_jones").password("open sesame")
				.roles("USER").and().withUser("superman").password("kryptonite").roles("ALIEN");
	}

	@Override
	protected void configure(HttpSecurity httpSecurity) throws Exception {

		httpSecurity.authorizeRequests().antMatchers("/createTreasure").hasRole("ALIEN").antMatchers("/findTreasure")
				.hasRole("USER").and().formLogin();

	}

	@Bean
	public PasswordEncoder getPasswordEncoder() {
		return NoOpPasswordEncoder.getInstance();
	}
}

```

## TreasureController .java
```java
@Controller
public class TreasureController {

	@RequestMapping("/")
	public @ResponseBody String welcome() {
		return "Welcome treasure hunter.";
	}

	@RequestMapping("/findTreasure")
	public @ResponseBody String findTreasure(@RequestParam String secretCode) {
		if ("open sesame".equals(secretCode)) {
			return "Congratulations! You have discovered the secret treasure.";
		}
		return "You are doomed!";
	}

	@RequestMapping("/createTreasure")
	public @ResponseBody String createTreasure() {
		return "A new hidden treasure has been created";
	}
}

```


## application.properties
```properties
#spring.security.user.name=indiana_jones
#spring.security.user.password=open sesame
```


![image](https://user-images.githubusercontent.com/13867858/168596307-98bde3ab-433b-4d7d-b46f-66b71784157d.png)
![image](https://user-images.githubusercontent.com/13867858/168596462-a05c991e-daf7-4388-b149-b1d539577d48.png)
![image](https://user-images.githubusercontent.com/13867858/168596705-b2678de5-ed4b-4ea9-8e03-6f8799edea75.png)
![image](https://user-images.githubusercontent.com/13867858/168599115-4c5eafad-11c5-4f25-888b-1d2cfcfd851f.png)
![image](https://user-images.githubusercontent.com/13867858/168599235-08b3f62e-9db5-42ca-832d-1b9af28151f7.png)





####
####
![image](https://user-images.githubusercontent.com/13867858/168580297-17a2c1b1-da4e-4bd9-afb8-9880d4dddea2.png)

![image](https://user-images.githubusercontent.com/13867858/168580389-2b758cb4-9856-43e7-a008-d3b28477cb5c.png)


![image](https://user-images.githubusercontent.com/13867858/168580482-8ba5f2a4-de96-4e64-80b3-de7995b1c2e3.png)
